packages/core/src/lib/code-snippet.svelte

```diff
@@ -0,0 +1,230 @@
+export { extraClasses as cx };
+
+let element: HTMLElement | undefined;
+
+$: copyButtonLabel = 'Copy';
+$: copyButtonIcon = 'content-copy' as IconName;
+
+const dispatch = createEventDispatcher<{
+  copy: { succeeded: boolean; message: string };
+}>();
+
+const copyToClipboard = async () => {
+  try {
+    await navigator.clipboard.writeText(code);
+
+    copyButtonLabel = 'Copied';
+    copyButtonIcon = 'check';
+
+    dispatch('copy', {
+      succeeded: true,
+      message: 'Successfully copied snippet to the clipboard',
+    });
+  } catch {
+    copyButtonLabel = 'Failed';
+    copyButtonIcon = 'alert';
+
+    dispatch('copy', {
+      succeeded: false,
+      message: 'Failed to copy snippet to the clipboard',
+    });
+  }
+
+  window.setTimeout(() => {
+    copyButtonLabel = 'Copy';
+    copyButtonIcon = 'content-copy';
+  }, 2000);
+};
+
+const highlight = () => {
+  if (element) {
+    Prism.highlightElement(element);
+  }
+};
+
+const formatCode = (input: string): string => {
+  const htmlEntities: Record<string, string> = {
+    '<': '&lt;',
+    '>': '&gt',
+    '/': '&#47;',
+  };
+
+  let formattedCode = input;
+  for (const [key, value] of Object.entries(htmlEntities)) {
+    if (formattedCode) {
+      formattedCode = formattedCode.replaceAll(key, value);
+    }
+  }
+
+  return formattedCode;
+};
+
+$: {
+  if (element && code) {
+    element.innerHTML = formatCode(code);
+    highlight();
+  }
+}
+
+onMount(async () => {
+  try {
+    /**
+     * Load the themes, for vite to properly discover them we cannot use string templates,
+     * so we use a switch statement for our supported themes.
+     */
+    switch (theme) {
+      case 'vs': {
+        await import('prism-themes/themes/prism-vs.min.css');
+        break;
+      }
+      case 'vsc-dark-plus': {
+        await import('prism-themes/themes/prism-vsc-dark-plus.min.css');
+        break;
+      }
+      default: {
+        // Unsupported theme
+        break;
+      }
+    }
+
+    /**
+     * After the HTML is loaded in the DOM, we can use the autoloader to manage
+     * scanning for languages and including the components properly
+     */
+    await import(
+      // @ts-expect-error no type declaration for this JS file
+      'prismjs/plugins/autoloader/prism-autoloader'
+    );
+
+    // Make sure the autoloader knows where to find our languages
+    (Prism.plugins.autoloader as { languages_path: string }).languages_path =
+      // TODO: do we want a better way to get the version?
+      'https://cdnjs.cloudflare.com/ajax/libs/prism/1.29.0/components/';
+
+    // Do the initial highlighting
+    if (element) {
+      Prism.highlightElement(element);
+    }
+  } catch (error) {
+    /* eslint-disable-next-line no-console */
+    console.error('Error initializing prismjs', error);
+  }
+});
+</script>
+
+<figure class={cx('flex flex-col gap-2', extraClasses)}>
+  {#if $$slots.default}
+    <figcaption><slot /></figcaption>
+  {/if}
```

Comment:
Added this to allow an easy way to accessibly caption the code-snippets. After talking to Nat, we would generally provide copy contextualizing the snippet above the code block, so this could be a useful way to associate that contextual copy with the `figure` itself. Open to not include this but seems like an easy feature to bake-in that could be handy.